datalad/support/sshconnector.py

```diff
@@ -450,7 +314,351 @@ def get_git_version(self):
+
+        Raises
+        ------
+        ConnectionOpenFailedError
+          When starting the SSH ControlMaster process failed.
+        """
+        # the socket should vanish almost instantly when the connection closes
+        # sending explicit 'check' commands to the control master is expensive
+        # (needs tempfile to shield stdin, Runner overhead, etc...)
+        # as we do not use any advanced features (forwarding, stop[ing the
+        # master without exiting) it should be relatively safe to just perform
+        # the much cheaper check of an existing control path
+        if self.ctrl_path.exists():
+            return False
+
+        # create ssh control master command
+        cmd = ["ssh"] + self._ssh_open_args + self._ssh_args + [self.sshri.as_str()]
+
+        # start control master:
+        lgr.debug("Opening %s by calling %s" % (self, cmd))
+        proc = Popen(cmd)
+        stdout, stderr = proc.communicate(input="\n")  # why the f.. this is necessary?
+
+        # wait till the command exits, connection is conclusively
+        # open or not at this point
+        exit_code = proc.wait()
+
+        if exit_code != 0:
+            raise ConnectionOpenFailedError(
+                cmd,
+                'Failed to open SSH connection (could not start ControlMaster process)',
+                exit_code,
+                stdout,
+                stderr,
+            )
+        self._opened_by_us = True
+        return True
+
+    def close(self):
+        """Closes the connection.
+        """
+        if not self._opened_by_us:
+            lgr.debug("Not closing %s since was not opened by itself", self)
+            return
+        # stop controlmaster:
+        cmd = ["ssh", "-O", "stop"] + self._ssh_args + [self.sshri.as_str()]
+        lgr.debug("Closing %s by calling %s", self, cmd)
+        try:
+            self._runner.run(cmd, expect_stderr=True, expect_fail=True)
+        except CommandError as e:
+            lgr.debug("Failed to run close command")
+            if self.ctrl_path.exists():
+                lgr.debug("Removing existing control path %s", self.ctrl_path)
+                # socket need to go in any case
+                self.ctrl_path.unlink()
+            if e.code != 255:
+                # not a "normal" SSH error
+                raise e
+
+
+@auto_repr
+class BaseSSHManager(object):
+    """Interface for an SSHManager
+    """
+    def ensure_initialized(self):
+        """Assures that manager is initialized"""
+        pass
+
+    assure_initialized = ensure_initialized
+
+    def get_connection(self, url, use_remote_annex_bundle=True, force_ip=False):
+        """Get an SSH connection handler
+
+        Parameters
+        ----------
+        url: str
+          ssh url
+        force_ip : {False, 4, 6}
+          Force the use of IPv4 or IPv6 addresses.
+
+        Returns
+        -------
+        BaseSSHConnection
+        """
+        raise NotImplementedError
```

Comment:
should we start using abc and `@abstractmethod` more?  not that I foresee other subclasses to come but IMHO getting consistent use of `abc` for interface/base classes would be of advantage

Reply:
I am not familiar with them. But maybe not in the context of an urgent fix